client/http/client.go

```diff
@@ -303,12 +305,17 @@ func NewClient(
 		return nil
 	}
 	c.inner.init(sd)
+	c.defaultSD = true
 	return c
 }
 
 // Close gracefully closes the HTTP client.
 func (c *client) Close() {
 	c.inner.close()
+	// only close the service discovery if it's created by the client.
+	if c.defaultSD && c.inner.sd != nil {
+		c.inner.sd.Close()
+	}
```

Comment:
What about wrapping the `defaultSD` into the inner client and providing a method to close `sd` if necessary?

Reply:
moved `defaultSD` into `inner`
I don't think it seems necessary to provide a function
- For default creation, there is no need to care about close sd
- For incoming sd creation, it should be left to the higher level to do the closing.